gtsfm/common/view_graph.py

```diff
@@ -0,0 +1,34 @@
+"""Defines the ViewGraph data structure.
+
+Authors: Akshay Krishnan
+"""
+from dataclasses import dataclass
+from typing import Dict, Tuple, List
+
+import numpy as np
+from gtsam import Rot3, Unit3, Cal3Bundler
+
+
+@dataclass
+class ViewGraph:
```

Comment:
I'm slightly apprehensive about calling this a graph without defining the actual structure of the graph a little more rigorously. 

For example, I'm assuming the vertices should be the individual cameras and should store the calibrations and absolute pose information (set to identity if global poses have not been computed). The edges indicate that the cameras observe enough of the same landmarks to compute their relative geometry, and the edges should store this information along with the correspondences. I think representing the edges as a single dictionary with image pair keys would be a lot cleaner, or at least having some (public) master list of image pairs that is then used to access the (private) dictionaries.

Reply:
> I think representing the edges as a single dictionary with image pair keys would be a lot cleaner

It would be cleaner, yes. But where is this structure for an edge actually used? The outputs of this class are fed to rotation and translation averaging which need dicts of rotations and translations (separately).  I am open to changing APIs for these classes but that needs more discussion and is beyond the scope of this PR. 

> or at least having some (public) master list of image pairs 

Yeah, we can add a member to store valid image pairs.

Reply:
Representing the ViewGraph as an actual graph is a useful data abstraction, especially when we want to start considering [more sophisticated view filtering algorithms](https://arxiv.org/pdf/1902.02256.pdf). I think what you've done so far is great, and I think a very minimal refactoring will make the graph structure more obvious. Here's a quick draft of what I had in mind (somewhat motivated by [DAGSfM](https://github.com/AIBluefisher/DAGSfM/blob/dev/src/graph/view_graph.h)):

```python
class TwoViewData(NamedTuple):
    i2Ti1: Pose3
    corr_idxs_i1i2: np.ndarray
   
class ViewGraph:
    # Put these in an __init__ and initialize from their constituent parts
    self.view_pairs: List[Tuple[int, int]]
    self._views: Dict[int, PinholeCameraCal3Bundler]
    self._view_edges: Dict[Tuple[int, int], TwoViewData]
    
    # Add getters for the attributes of TwoViewData.
    def get_i2Ui1(i1, i2):
        return self._view_edges[(i1, i2)].i2Ti1.translation()

    # etc., etc.
```

Reply:
> It would be cleaner, yes. But where is this structure for an edge actually used? The outputs of this class are fed to rotation and translation averaging which need dicts of rotations and translations (separately). I am open to changing APIs for these classes but that needs more discussion and is beyond the scope of this PR.

The edge is just holds any relative view information, e.g., relative translation and rotations (which you could combine into a pose or keep separate), two view correspondences, etc.

Reply:
> It does not store the absolute pose information. I am open to other names if you think it is misleading. Maybe TwoViewGraph. The definition is clear to me: it is a structure that stores all the two-view relationships before averaging (including intrinsic priors). I can define an edge in the documentation if that is what your comment is about.

I think having the absolute information would be useful. Then, we could pass the ViewGraph to averaging, and it would output a new ViewGraph with the absolute pose data filled in.

Reply:
I understood the design you had in mind from your previous comments, but I already mentioned my thoughts on it: 
- We cannot have PinholeCameraCal3Bundler as we do not have the absolute camera pose from two-view information. 
- I like the TwoViewData struct, but to pass inputs to RA and TA we would have to re-extract a Dict[Tuple[int, int], Rot3] and Dict[Tuple[int, int], Unit3], which is unnecessary. I think changes to RA and TA should be done in a different PR. 

The edge is holding two-view information even now. The question is how should this be internally represented. Multiple dicts vs single dict. I would prefer a representation that avoids redundant computation.

Reply:
> We cannot have PinholeCameraCal3Bundler as we do not have the absolute camera pose from two-view information.

I think this class could useful even after RA and TA, so I think at least having the option to populate the absolute poses later would be a good idea.

> I like the TwoViewData struct, but to pass inputs to RA and TA we would have to re-extract a Dict[Tuple[int, int], Rot3] and Dict[Tuple[int, int], Unit3], which is unnecessary. I think changes to RA and TA should be done in a different PR.

Sure, then just forego the `TwoViewData` for now and just have something like:

```python
class ViewGraph:
    # Put these in an __init__ and initialize from their constituent parts
    self.view_pairs: List[Tuple[int, int]]
    self._views: Dict[int, PinholeCameraCal3Bundler]
    self._i1Ri2: Dict[Tuple[int, int], Rot3]
    self._i1ti2: Dict[Tuple[int, int], Union[Unit3, Point3]]
    self._corr_idxs_i1i2: Dict[Tuple[int, int], np.ndarray]
    # optionally add keypoints?
    
    # Add getters for the attributes of TwoViewData.
    def get_i2Ui1(i1, i2):
        return self._view_edges[(i1, i2)].i2Ti1.translation()

    # etc., etc.
```

Reply:
Sure, although I would stick with Cal3Bundler and introduce PinholeCamera only in a PR where its populated.

Reply:
That works for me. 

Would you mind creating an issue outlining that we'd eventually like to pass a ViewGraph to the averaging module (which would then populate the absolute poses)?

Reply:
I am not convinced about this design of having an incomplete data structure, and one-by-one modules filling up more information in the same structure.

Are there any advantage of this approach? Having attributes which can be `None` is something I try to avoid.

What we want in this PR is a way to capture two-view relationships (edges) between images. And we want to just keep the edges which are good 2-view relationship. Having absolute poses here just muddies what we introduce as the output to the `ViewGraphEstimator` module. Having the estimator output an object with missing attributes do not make sense to me. 

Also, I agree with Akshay that we should just pass as inputs to a module which is needed for computation. As we follow a functional pattern for gtsfm, passing in objects with `None` attributes and filling it in the module is not something we would do. We would make a copy of the whole data-structure. I find this needless.

I would infact advocate to remove calibrations from the dataclass too, and have just edge information there. No need to put any property of the node in there.

Reply:
> What we want in this PR is a way to capture two-view relationships (edges) between images. And we want to just keep the edges which are good 2-view relationship. Having absolute poses here just muddies what we introduce as the output to the ViewGraphEstimator module. Having the estimator output an object with missing attributes do not make sense to me.

Edges encode correspondences between images, and correspondences can be filtered using the absolute poses, e.g., what we're already doing in data association.

> Also, I agree with Akshay that we should just pass as inputs to a module which is needed for computation. As we follow a functional pattern for gtsfm, passing in objects with None attributes and filling it in the module is not something we would do. We would make a copy of the whole data-structure. I find this needless.

It doesn't have to be None; it could just be identity. Also, `GtsfmData` is "filled in" after initialization, and there are plenty of other examples of classes with Optional attributes.

> I would infact advocate to remove calibrations from the dataclass too, and have just edge information there. No need to put any property of the node in there.

If the nodes are inconsequential, then this isn't really a graph.

Reply:
> I would infact advocate to remove calibrations from the dataclass too, and have just edge information there. No need to put any property of the node in there.

@ayushbaid I would be okay with removing calibrations for now as RCC does not use it. We can introduce it in the same PR as the fundamental matrix based cycle-consistency if that works well.